### ISSUE TITLE
native docker command and docker compose works differently

### ISSUE THREAD

**Issue by lincze** [vfarcic/books-ms#1]
I've made a fresh installation, and (surprise) the native docker command

```
sudo docker run -it --rm \
-v $PWD/client/components/:/source/client/components \
-v $PWD/client/test:/source/client/test \
-v $PWD/src:/source/src \
-v $PWD/target:/source/target \
-p 8080:8080 \
--env TEST_TYPE=watch-front vfarcic/books-ms-tests
```

and the docker-compose command

```
sudo docker-compose -f docker-compose-dev.yml run feTestsLocal
```

works differently. The 1st one, as expected, but the second one gives an error:

```
404 GET /api/v1/books
firefox 43               ✖ test/tc-books.html » "after each" hook

  Error: Anonymous mock already called once (http://localhost:2000/components/sinonjs/sinon.js:4008)
    <unknown> at           .fire at /components/client/bower_components/polymer/polymer.html.js:1317:0
    <unknown> at   ._handleError at /components/client/bower_components/iron-ajax/iron-ajax.html.js:431:0

404 GET /api/v1/books
firefox 43               Tests failed: 1 failed tests
Test run ended in failure: 1 failed tests

firefox 43 (85/0/1)                   
[04:43:56] 'wct:local' errored after 1.68 min
[04:43:56] Error: 1 failed tests
```

really can't imagine, why.

```
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ docker --version
Docker version 1.10.1, build 9e83765
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ docker-compose --version
docker-compose version: 1.4.1
```

last commit:

```
commit ca3e29704a87ce03fbc96bd8a471fdf951b5ad03
Author: vfarcic <viktor@farcic.com>
Date:   Fri Jan 29 00:04:08 2016 +0100

    Added net
```

**Comment by lincze** [comment]
I've upgraded the docker-compose to 1.6.0 (which is the release brother of the docker engine installed with the apt. The docker-compose command

```
sudo docker-compose -f docker-compose-dev.yml run feTestsLocal
```

then worked as expected. To be sure that this is it, what takes the difference, I've ran the command with the 1.4.1 docker-compose again and got in fact the error mentioned in the previous post. There are subtle differencies between the two runs before the error (e.g. compose 1.6.0 starts with a message: "WARNING: The DOMAIN variable is not set. Defaulting to a blank string.")

Not a good news, if 1.4.1 compose is not fully usable with engine 1.10.1.

**Comment by lincze** [comment]
Probably, the best fix for this issue to fix the docker engine, compose, etc. version numbers to values/ranges known to work together. For me, the 1.10.1 engine (which was given by 'latest') works with 1.6.0 compose

**Comment by vfarcic** [comment]
Sorry for not responding earlier. For some unknown reason I didn't get a notification about this issue.

I don't think that the problem is related with Docker or Docker Compose versions but with tests running in parallel. The problem appears only very occasionally and rerunning the command fixes it.

The WARNING message is OK. There is an environment variable inside the docker-compose.yml. When not set, it defaults to an empty string.

**Comment by lincze** [comment]
ok, though i have ecperienced (in general, not with your book) slight differencies between versions, problems at upgrades, etc. so, all in all, at a component changing so rapidly as docker, i would completeley avoid the ":latest" version. that's all (and i really love your book).

**Comment by vfarcic** [comment]
I had an internal doubt what to do with versions. On one hand, if I version everything, the examples will be more stable. On the other hand, the code is always using the latest, people will be able to experience the current tools instead being locked in something that is already outdated. I choose later precisely because things are changing too fast. At the same time, I am receiving feedback and correcting problems as soon as I can. For example, Ansible had problems in the last two releases before 2.0 so I changed from latest to a fixed version (I don't remember which one it was). Later on, when Ansible 2.0 was released and those problems disappeared, I changed it back to latest. The similar thing happened with Docker Compose. Until few days ago, a fixed version was used due to few problems. The last push to the repository has the Ansible role set to use the latest Docker Compose.

I think that both approaches have pros and cons and I'm still not 100% sure which one is better in this case.
